server/src/main/java/io/crate/planner/optimizer/joinorder/JoinGraph.java

```diff
@@ -0,0 +1,317 @@
+
+    JoinGraph joinWith(LogicalPlan root, JoinGraph other) {
+        for (var node : other.nodes) {
+            assert !edges.containsKey(node) : "LogicalPlan" + node + " can't be in both graphs";
+        }
+
+        var newNodes = Lists2.concat(this.nodes, other.nodes);
+        var newEdges = Maps.merge(this.edges, other.edges, Sets::union);
+        var newFilters = Lists2.concat(this.filters, other.filters);
+        var hasCrossJoin = this.hasCrossJoin || other.hasCrossJoin();
+
+        return new JoinGraph(
+            root,
+            newNodes,
+            newEdges,
+            newFilters,
+            hasCrossJoin
+        );
+    }
+
+    JoinGraph withEdges(Map<LogicalPlan, Set<Edge>> edges) {
+        var newEdges = Maps.merge(this.edges, edges, Sets::union);
+        return new JoinGraph(this.root, this.nodes, newEdges, this.filters, this.hasCrossJoin);
+    }
+
+    JoinGraph withFilters(List<Symbol> filters) {
+        var newFilters = Lists2.concat(this.filters, filters);
+        return new JoinGraph(this.root, this.nodes, edges, newFilters, this.hasCrossJoin);
+    }
+
+    JoinGraph withCrossJoin(boolean hasCrossJoin) {
+        return new JoinGraph(this.root, this.nodes, edges, filters, hasCrossJoin);
+    }
+
+    public int size() {
+        return nodes.size();
+    }
+
+    public Set<Edge> edges(LogicalPlan node) {
+        var result = edges.get(node);
+        if (result == null) {
+            return Set.of();
+        }
+        return result;
+    }
+
+    public LogicalPlan buildLogicalPlan() {
+        return reorder(this.nodes);
+    }
+
+    public LogicalPlan reorder(List<LogicalPlan> order) {
+        var result = order.get(0);
+        var alreadyJoinedNodes = new HashSet<LogicalPlan>();
+        alreadyJoinedNodes.add(result);
+
+        for (int i = 1; i < order.size(); i++) {
+            var rightNode = order.get(i);
+            alreadyJoinedNodes.add(rightNode);
+
+            var criteria = new ArrayList<Symbol>();
+
+            for (var edge : this.edges(rightNode)) {
+                var toNode = edge.to();
+                if (alreadyJoinedNodes.contains(toNode)) {
+                    criteria.add(EqOperator.of(edge.left(), edge.right()));
+                }
+            }
+
+            if (criteria.isEmpty()) {
+                var errorMessage = new ArrayList<String>();
+                for (var plan : order) {
+                    for (var relationName : plan.getRelationNames()) {
+                        errorMessage.add(relationName.fqn());
+                    }
+                }
+                throw new InvalidArgumentException("JoinPlan cannot be built with the provided order " + errorMessage);
+            }
+
+            result = new JoinPlan(
+                result,
+                rightNode,
+                JoinType.INNER,
+                AndOperator.join(criteria, null),
+                false,
+                true
```

Comment:
question: could you explain how `crossJoinEliminationRulePassed` can be `true` here?

Reply:
You are right, the name of the property and the location is confusing. This should be set in the eliminate cross-join rule and probably better called isReordered.

Reply:
I think I will the reorder method from JoinGraph into the rule because it makes logically more sense.